microsoft/botbuilder-dotnet
# Authentication: Add opt-in support for MSAL to acquire tokens

Fixes #5735

# Additive, opt-in change
Adds `MsalAppCredentials` which allows to use MSAL instead of ADAL. `MsalAppCredentials` collapses the equivalent of `MicrosoftAppCredentials` plus `AdalAuthenticator`, except that it uses MSAL and has less complexity.

The change does *not* affect customers by default, instead, it works as an opt-in mechanism. Also, the factory is very open ended, depending on the MSAL `IConfidentialClientApplication` which allows DI friendly registration of cert, password or other authentication mechanisms without any coupling on our side.

Example of opt-in startup.cs when using adaptive runtime for password:

```csharp
            // MSAL secret auth:
            services.AddSingleton<IConfidentialClientApplication>(
                serviceProvider => ConfidentialClientApplicationBuilder.Create(Configuration.GetSection("MicrosoftAppId").Value)
                    .WithClientSecret(Configuration.GetSection("MicrosoftAppPassword").Value)
                    .Build());

            // MSAL credential factory: regardless of secret, cert or custom auth, need to add the line below
            // to enable MSAL
            services.AddSingleton<ServiceClientCredentialsFactory, MsalServiceClientCredentialsFactory>();
```

Example of opt-in startup.cs when using adaptive runtime for certificate auth:

```csharp
            // MSAL certificate auth:
            services.AddSingleton<IConfidentialClientApplication>(
                serviceProvider => ConfidentialClientApplicationBuilder.Create(Configuration.GetSection("MicrosoftAppId").Value)
                    .WithCertificate(new X509Certificate2("<path to cert file>", "certPassword"))
                    .Build());

            // MSAL credential factory: regardless of secret, cert or custom auth, need to add the line below
            // to enable MSAL
             services.AddSingleton<ServiceClientCredentialsFactory, MsalServiceClientCredentialsFactory>();
```

In scenarios that don't use the new adaptive runtime, and instead use the AddBot<IBot> method, the pattern is to create an instance of `MsalAppCredentials` with the preferred authentication type, and then pass it as part of the `BotFrameworkOptions`.

# Testing

- Manual testing with skills
- Integration tests (need to be ported to Functional-tests repo)
- Parallel tests and measurements

During parallel tests, we re-verified the investigation done when writing `AdalAuthenticator`, obtaining similar results. When we don't control concurrency, when the token expires we get huge contention, timeouts even with retries and waits of 20+ seconds (to try for yourself, remove the semaphore and run the parallel load test provided in this PR). With the semaphore we observe no contention and fast recovery after token acquisition even under high parallel load.

# Next steps

We'll bake this as opt-in post R14, and in R15, once this has been exercised further, we'll make it the default auth pattern.

Another long term TODO is to remove our custom retry and use Polly.NET for all retriable operations (should likely be done along with an all up analysis of what should and should not be retried and how to allow developers control key aspects of that).

# V5 impact

In V5 also we should consider deprecating all *Appcredentials classes, including MSAL AppCredentials and just expose the new factory pattern.
